x/feedistribution/keeper/allocation.go

```diff
@@ -0,0 +1,92 @@
+package keeper
+
+import (
+	"cosmossdk.io/math"
+
+	"github.com/ExocoreNetwork/exocore/x/feedistribution/types"
+	cryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+// Based on the epoch, AllocateTokens performs reward and fee distribution to all validators.
+func (k Keeper) AllocateTokens(ctx sdk.Context, totalPreviousPower int64) error {
+	feeCollector := k.authKeeper.GetModuleAccount(ctx, k.feeCollectorName)
+	feesCollectedInt := k.bankKeeper.GetAllBalances(ctx, feeCollector.GetAddress())
+	feesCollected := sdk.NewDecCoinsFromCoins(feesCollectedInt...)
+
+	// transfer collected fees to the distribution module account
+	if err := k.bankKeeper.SendCoinsFromModuleToModule(ctx, k.feeCollectorName, types.ModuleName, feesCollectedInt); err != nil {
+		return err
+	}
+
+	feePool := k.GetFeePool(ctx)
+	if totalPreviousPower == 0 {
+		feePool.CommunityPool = feePool.CommunityPool.Add(feesCollected...)
+		k.SetFeePool(ctx, feePool)
+		return nil
+	}
+
+	// calculate fraction allocated to exocore validators
+	remaining := feesCollected
+	communityTax, err := k.GetCommunityTax(ctx)
+	if err != nil {
+		return err
+	}
+	feeMultiplier := feesCollected.MulDecTruncate(math.LegacyOneDec().Sub(communityTax))
+
+	// allocate tokens proportionally to voting power of different validators
+	//
+	// TODO: Consider parallelizing later
+	validatorUpdates := k.StakingKeeper.GetValidatorUpdates(ctx)
```

Comment:
Isn't this the `new` voting power? And you're using it to allocate past rewards?